stream/stream_dvbv5.c

```diff
@@ -0,0 +1,596 @@
+    priv->opts = priv->opts_cache->opts;
+
+    stream->fill_buffer = dvbv5_streaming_read;
+    stream->close = dvbv5_close;
+    stream->control = dvbv5_stream_control;
+    stream->streaming = true;
+    stream->demuxer = "lavf";
+    stream->lavf_type = "mpegts";
+
+    if (priv->opts->cfg_lnb_name) {
+        lnb = dvb_sat_search_lnb(priv->opts->cfg_lnb_name);
+    }
+
+    dvb = dvb_dev_alloc();
+    if (!dvb)
+        return STREAM_ERROR;
+
+    if (priv->opts->cfg_server && priv->opts->cfg_port) {
+        MP_INFO(stream, "Connecting to %s:%d\n", priv->opts->cfg_server, priv->opts->cfg_port);
+        ret = dvb_dev_remote_init(dvb, priv->opts->cfg_server, priv->opts->cfg_port);
+        if (ret < 0) {
+            MP_ERR(stream, "dvb remote init failed: %i\n", ret);
+            return STREAM_ERROR;
+        }
+    }
+
+    dvb_dev_set_log(dvb, priv->opts->cfg_verbose, NULL);
+    dvb_dev_find(dvb, NULL, NULL);
+    parms = dvb->fe_parms;
+
+    dvb_dev = dvb_dev_seek_by_adapter(dvb, priv->opts->cfg_adapter, priv->opts->cfg_demux, DVB_DEVICE_DEMUX);
+    if (!dvb_dev) {
+        dvb_dev_free(dvb);
+        return STREAM_ERROR;
+    }
+
+    demux_dev = dvb_dev->sysname;
+
+    dvb_dev = dvb_dev_seek_by_adapter(dvb, priv->opts->cfg_adapter, priv->opts->cfg_demux, DVB_DEVICE_DVR);
+    if (!dvb_dev) {
+        MP_ERR(stream, "Couldn't find dvr device node\n");
+        dvb_dev_free(dvb);
+        return STREAM_ERROR;
+    }
+    dvr_dev = dvb_dev->sysname;
+
+    dvb_dev = dvb_dev_seek_by_adapter(dvb, priv->opts->cfg_adapter, priv->opts->cfg_frontend, DVB_DEVICE_FRONTEND);
+    if (!dvb_dev)
+        return STREAM_ERROR;
+
+    if (!dvb_dev_open(dvb, dvb_dev->sysname, O_RDWR)) {
+        dvb_dev_free(dvb);
+        return STREAM_ERROR;
+    }
+    if (lnb >= 0)
+        parms->lnb = dvb_sat_get_lnb(lnb);
+    if (priv->opts->cfg_sat_number >= 0)
+        parms->sat_number = priv->opts->cfg_sat_number;
+    parms->diseqc_wait = priv->opts->cfg_diseqc_wait;
+    parms->freq_bpf = priv->opts->cfg_freq_bpf;
+    parms->lna = priv->opts->cfg_lna;
+
+    r = dvb_fe_set_default_country(parms, priv->opts->cfg_cc);
+    if (r < 0)
+        MP_ERR(stream, "Failed to set the country code:%s\n", priv->opts->cfg_cc);
+
+    if (parse(stream, parms, &vpid, &apid, &sid)) {
+        dvb_dev_free(dvb);
+        return STREAM_ERROR;
+    }
+
+    if (setup_frontend(stream, parms) < 0) {
+        dvb_dev_free(dvb);
+        return STREAM_ERROR;
+    }
+
+    if (priv->opts->cfg_rec_psi) {
+        if (sid < 0) {
+            MP_ERR(stream, "Service id 0x%04x was not specified at the file\n",
+                sid);
+            dvb_dev_free(dvb);
+            return STREAM_ERROR;
+        }
+
+        sid_fd = dvb_dev_open(dvb, demux_dev, O_RDWR);
+        if (!sid_fd) {
+            MP_ERR(stream, "opening sid demux failed\n");
+            dvb_dev_free(dvb);
+            return STREAM_ERROR;
+        }
+        pmtpid = dvb_dev_dmx_get_pmt_pid(sid_fd, sid);
```

Comment:
This function call never returns for me (0 CPU load) when using a DVB-S/S2 card, but that seems like something in `libdvbv5` getting stuck. Will have a look later.

Reply:
Ok, also this freezes for me:
```
dvbv5-zap -v -c ~/.mpv/dvb_channel.conf -p "Das Erste HD"
```
Adding a `sleep` before calling `dvb_dev_dmx_get_pmt_pid` "fixes" it, so it seems to be a weird timing issue.

Reply:
The cleanest fix seems to be to move `check_frontend` (and the thread doing the timeout check) right after:
```
    if (setup_frontend(stream, parms) < 0) {
        dvb_dev_free(dvb);
        return STREAM_ERROR;
    }
```
to ensure the channel is locked before trying to read data from it. I'd say this is a bug also in `dvbv5-zap`, reading data from the device before ensuring the channel is locked seems to get stuck sometimes. Do you want to report it upstream?

Reply:
Given:
https://git.linuxtv.org/v4l-utils.git/commit/?id=cfdaff0fe2f11ede4ac57d600d95b2ff945239ca
I think activating `dvbv5-pat` might be best default-on for a video player (after fixing the "getting stuck" issue).